File: bigseqkit/rename.go
package bigseqkit

import (
	"ignis/driver/api"
	"ignis/executor/api/ipair"
)

type SeqKitRenameOptions struct {
	inner RenameOptions
}

type RenameOptions struct {
	Config KitConfig
	ByName *bool
}

func (this *RenameOptions) setDefaults() *RenameOptions {
	this.Config.setDefaults()
	setDefault(&this.ByName, false)

	return this
}

func (this *SeqKitRenameOptions) Config(v *SeqKitConfig) *SeqKitRenameOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitRenameOptions) ByName(v bool) *SeqKitRenameOptions {
	this.inner.ByName = &v
	return this
}

func Rename(input *api.IDataFrame[string], o *SeqKitRenameOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitRenameOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	libprepare, err := api.AddParam(libSource("RenamePrepare"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	prepared, err := api.MapPartitions[string, ipair.IPair[string, string]](input, libprepare)
	if err != nil {
		return nil, err
	}

	ready, err := api.GroupByKey[string, string](api.ToPair[string, string](prepared), nil)
	if err != nil {
		return nil, err
	}

	librename, err := api.AddParam(libSource("Rename"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	return api.Flatmap[ipair.IPair[string, []string], string](ready.FromPair(), librename)
}

File: bigseqkit-py/bigseqkit/rename.py
from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitRenameOptions:

    def __init__(self):
        self.__inner = RenameOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def byName(self, v: bool):
        self.__inner.ByName = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()

        libprepare = _libSource("RenamePrepare").addParam("opts", _optionsToString(opts))
        prepared = input.mapPartitions(libprepare)

        ready = prepared.toPair().groupByKey()

        librename= _libSource("Rename").addParam("opts", _optionsToString(opts))
        return ready.flatmap(librename)

class RenameOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.ByName = None  # bool

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "ByName", False)


def rename(input: IDataFrame, o: SeqKitRenameOptions = None, **kwargs):
    if o is None:
        o = SeqKitRenameOptions()
    return o._run(input, **kwargs)
